File: app/src/main/java/com/tobikster/medicreminder/MedicReminderApplication.java
package com.tobikster.medicreminder;

import android.app.Activity;

import com.squareup.leakcanary.LeakCanary;
import com.tobikster.medicreminder.ui.RemindersListActivity;

import net.danlew.android.joda.JodaTimeAndroid;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;
import dagger.android.support.DaggerApplication;
import timber.log.Timber;

public class MedicReminderApplication extends DaggerApplication {
	@Inject
	DispatchingAndroidInjector<Activity> activityDispatchingAndroidInjector;

	@Override
	public void onCreate() {
		super.onCreate();

		if (LeakCanary.isInAnalyzerProcess(this)) {
			// This process is dedicated to LeakCanary for heap analysis.
			// You should not init your app in this process.
			return;
		}

		if (BuildConfig.DEBUG) {
			Timber.plant(new Timber.DebugTree());
		}

		LeakCanary.install(this);

		JodaTimeAndroid.init(this);
	}

	@Override
	protected AndroidInjector<? extends DaggerApplication> applicationInjector() {
		return DaggerMedicReminderApplication_Component.builder().create(this);
	}

	@dagger.Component(modules = {AndroidSupportInjectionModule.class, RemindersListActivity.Module.class})
	interface Component extends AndroidInjector<MedicReminderApplication> {
		@dagger.Component.Builder
		abstract class Builder extends AndroidInjector.Builder<MedicReminderApplication> {}
	}
}

File: app/src/main/java/com/tobikster/medicreminder/MedicReminderApplication.kt
package com.tobikster.medicreminder

import android.app.Activity
import android.app.Application
import com.squareup.leakcanary.LeakCanary
import com.tobikster.medicreminder.di.DaggerAppComponent
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasActivityInjector
import timber.log.Timber
import javax.inject.Inject

class MedicReminderApplication : Application(), HasActivityInjector {
	@Inject lateinit var activityDispatchingAndroidInjector: DispatchingAndroidInjector<Activity>

	override fun onCreate() {
		super.onCreate()

		if (LeakCanary.isInAnalyzerProcess(this)) {
			// This process is dedicated to LeakCanary for heap analysis.
			// You should not init your app in this process.
			return
		}

		DaggerAppComponent.builder().application(this).build().inject(this)

		if (BuildConfig.DEBUG) {
			Timber.plant(Timber.DebugTree())
		}

		LeakCanary.install(this)
	}


	override fun activityInjector(): AndroidInjector<Activity>? {
		return activityDispatchingAndroidInjector
	}
}
